Community live updates

Enderedne opened the pull request:
[WIP] Adjust messaging to support subscriptions

### Describe the background of your pull request

- need to implement graphql guard for subscriptions
- need to work on the matrix client sync state
- message event is fired N times for some reason

### Additional context

- [x] fix room/group invitations
- [x] Dealing with unlinked communities
- [x] pool mgmt / desposing of agents

Enderedne committed:
[WIP] Adjust messaging to support subscriptions
- fix room/group invitations
- need to work on the matrix client sync state
- message event is fired N times for some reason

techsmyth committed:
added descriptions

techsmyth committed:
prototype of working subscriptions with applications

techsmyth committed:
demonstrator using existing authentication / authorization for subscriptions

Enderedne committed:
Merge pull request #1357 from alkem-io/subs-auth

Subs authentication / authorisation setup

techsmyth committed:
added auth to msg subs

Enderedne committed:
Fix race condition during matrix state sync

Enderedne committed:
Expose community information
- membership data now includes community info
- communityId is stored on the matrix room-accountdata under the admin account
- expose communityId through subscription

Enderedne committed:
Fix room account data persistance

techsmyth committed:
Merge branch 'develop' into server-937

techsmyth commented on src/app.module.ts:
```diff
@@ -99,10 +99,25 @@ import { SsiAgentModule } from './services/platform/ssi/agent/ssi.agent.module';
+          // Details in https://github.com/nestjs/docs.nestjs.com/issues/394
+          return { headers: { authorization: `${authHeader}` } };
+        },
+        onDisconnect: async (_: any, __: any) => {
```
Captured this for a later PR

techsmyth commented on src/domain/community/community/community.resolver.subscriptions.ts:
```diff
@@ -0,0 +1,62 @@
+import { Inject, LoggerService, UseGuards } from '@nestjs/common';
+import { Args, Resolver, Subscription } from '@nestjs/graphql';
+import { AuthorizationEngineService } from '@services/platform/authorization-engine/authorization-engine.service';
+import { APPLICATION_RECEIVED } from '@services/platform/subscription/subscription.events';
```
The set of event types should be an Enum

techsmyth commented on src/domain/community/user/user.resolver.subscriptions.ts:
```diff
@@ -8,20 +13,28 @@ import {
       value: CommunicationMessageReceived
     ) {
+      // Use this to update the sender identifer
+      // Todo: should not be doing any heavy work during the resolving
```
Agreed, put a note in the parent epic

techsmyth commented on src/domain/community/user/user.resolver.subscriptions.ts:
```diff
@@ -30,12 +43,30 @@ export class UserResolverSubscriptions {
+    await this.authorizationEngine.grantAccessOrFail(
+      agentInfo,
+      user.authorization,
+      AuthorizationPrivilege.READ,
```
Actually this should probably be the UPDATE privilege, which only the user has for themselves (or global admin etc)

techsmyth commented on src/services/domain/membership/membership.dto.user.result.entry.ecoverse.ts:
```diff
@@ -23,13 +24,24 @@ export class MembershipUserResultEntryEcoverse extends MembershipResultEntry {
+  @Field(() => MembershipCommunityResultEntry, {
+    description: 'The community that represents this ecoverse',
+  })
+  community: MembershipCommunityResultEntry;
```
Do not think this is needed also in the ecoverse entry given there is the top level communities array.

techsmyth commented on src/services/platform/communication/communication.service.ts:
```diff
@@ -69,7 +72,9 @@ export class CommunicationService {
-      sendMsgData.sendingUserEmail
+      sendMsgData.sendingUserEmail,
+      undefined,
+      await this.getNonElevatedMiddlewares()
```
Not sure I understand the naming of this method, can you please clarify?

techsmyth commented on src/services/platform/communication/communication.service.ts:
```diff
@@ -260,7 +292,11 @@ export class CommunicationService {
-    const matrixAgent = await this.matrixAgentPool.acquire(currentUserEmail);
+    const matrixAgent = await this.matrixAgentPool.acquire(
+      currentUserEmail,
+      undefined,
```
A quick scan shows that this is always "undefined"?

techsmyth commented on src/services/platform/matrix/agent/matrix.agent.ts:
```diff
@@ -20,6 +20,13 @@ import {
 } from '@services/platform/subscription/subscription.events';
+import { CommunicationMessageReceived } from '@services/platform/communication/communication.dto.message.received';
+
+export type MatrixAgentMiddlewares = {
```
Can you explain this new type + why it is needed?

techsmyth commented on src/services/platform/matrix/management/matrix.user.management.service.ts:
```diff
@@ -155,10 +155,10 @@ export class MatrixUserManagementService {
-        { username },
-        `Registration check failure: ${error}`
-      );
+      // this.matrixUserAdapterService.logMatrixUser(
```
Commented out code can probably just be removed

techsmyth commented on src/services/platform/subscription/subscription.events.ts:
```diff
@@ -1,3 +1,4 @@
 export const COMMUNICATION_MESSAGE_RECEIVED = 'communicationMessageReceived';
```
Move all these to be an Enum

techsmyth reviewed (changes requested):
Nice set of updates, keen to get this down to developer. Some tidy ups still needed, as per discussion / comments.

Enderedne committed:
Implement basic user caching

Enderedne committed:
Implement caching for matrix agent pools

Enderedne committed:
Fix matrix cache expiration

Enderedne committed:
Use user caching only for matrix related calls

Enderedne committed:
Use enums for named subscription events

Enderedne committed:
Drop not used community field
- no need to expose the membership ecoverse-community relationship at this point

Enderedne committed:
Use the room public name instead
- the hierarchy needs to be adjusted in order to read the communityId from a matrix room entity

Enderedne committed:
Ensure the user have accept room/group invites
- don't rely solely on matrix state sync

Enderedne committed:
Merge branch 'develop' into server-937

Enderedne committed:
Add mutation for community message removal
- sketch message editing

Enderedne committed:
Return messageId when removing
- add example mutation

techsmyth commented on graphql-samples/mutations/message/remove-update-community:
```diff
@@ -0,0 +1,11 @@
+mutation removeUpdateCommunity($msgData: CommunityRemoveMessageInput!) {
```
This should be "removeUpdateMessageFromCommunity"

techsmyth commented on src/domain/community/community/community.resolver.mutations.ts:
```diff
@@ -251,6 +261,33 @@ export class CommunityResolverMutations {
+    description: 'Removes an update message from the specified community',
+  })
+  @Profiling.api
+  async removeUpdateCommunity(
```
removeUpdateMessageFromCommunity

techsmyth commented on src/domain/community/community/community.resolver.mutations.ts:
```diff
@@ -275,4 +312,31 @@ export class CommunityResolverMutations {
+    description: 'Removes a discussion message from the specified community',
+  })
+  @Profiling.api
+  async removeDiscussionCommunity(
```
removeDiscussionMessageFromCommunity

techsmyth commented on src/services/platform/communication/communication.dto.delete.msg.community.ts:
```diff
@@ -0,0 +1,5 @@
+export class CommunicationDeleteMessageCommunityInput {
```
CommunicationDeleteMessageFromCommunityRoomInput - very long but very literal

techsmyth commented on src/services/platform/communication/communication.dto.edit.msg.community.ts:
```diff
@@ -0,0 +1,6 @@
+export class CommunicationEditMessageCommunityInput {
```
ditto

techsmyth commented on src/services/platform/communication/communication.service.ts:
```diff
@@ -82,6 +84,50 @@ export class CommunicationService {
     return messageId;
   }
 
+  async editMsgCommunity(
```
editMessageOnCommunityRoom

techsmyth commented on src/services/platform/subscription/subscription.events.ts:
```diff
@@ -1,3 +1,6 @@
-export const COMMUNICATION_MESSAGE_RECEIVED = 'communicationMessageReceived';
-export const MATRIX_ROOM_JOINED = 'matrixRoomJoined';
-export const USER_AVATAR_UPLOADED = 'avatarUploaded';
+export enum SubscriptionEvents {
```
Should be singular, and also up in the common/enums folder

techsmyth commented on src/services/platform/subscription/subscription.events.ts:
```diff
@@ -1,3 +1,6 @@
+export enum SubscriptionEvents {
+  COMMUNICATION_MESSAGE_RECEIVED = 'communicationMessageReceived',
+  APPLICATION_RECEIVED = 'applicationReceived',
+  MATRIX_ROOM_JOINED = 'matrixRoomJoined',
```
COMMUNICATION_ROOM_JOINED = ' communicationRoomJoined'

techsmyth reviewed (changes requested):
Looks nice. A few naming updates suggested which are all minor enough. Key question for Valentin that I have is related to the User cache, would like another look at that.

techsmyth committed:
addressed renaming feedback on PR

techsmyth reviewed (approved):
Feedback is addressed from my point of view. Over to Valentin...

valentinyanakiev committed:
Merged develop into server-937

valentinyanakiev committed:
Replaced all PubSub implementation to work against PubSubEngine (contract vs implementation)

valentinyanakiev commented on src/domain/community/user/user.module.ts:
```diff
@@ -25,6 +25,7 @@ import { SubscriptionModule } from '@services/platform/subscription/subscription
     AuthorizationEngineModule,
     TypeOrmModule.forFeature([User]),
     SubscriptionModule,
+    CacheModule.register({ max: 250, ttl: 300 }),
```
The built-in cache module in nestJS is in-memory caching. It is not horizontally scalable - we need to work against a contract and implementation like Redis.

techsmyth replied:
As discussed this is in-memory caching; caching in general to be looked at separately. What is there is good for now.

valentinyanakiev commented on src/domain/community/community/community.service.ts:
```diff
@@ -456,4 +459,38 @@ export class CommunityService {
+      community.discussionRoomID,
+      email
+    );
+    await this.communicationService.deleteMsgCommunity({
```
The methods and payloads in communicationService are inconsistent - e.g. deleteMsgCommunity, msgData, sendMsgCommunity. Make them consistent with the rest of the application - don't eat up words and don't use acronyms, pseudonyms etc.

techsmyth replied:
Mutations / methods / dtos are all renamed

valentinyanakiev commented on src/domain/community/community/community.resolver.subscriptions.ts:
```diff
@@ -0,0 +1,64 @@
+      AuthorizationPrivilege.UPDATE,
+      `subscribe to application events on community: ${community.displayName}`
+    );
+    return this.pubSub.asyncIterator(
```
This (and overall all events that should clearly be managed via lifecycle) should be managed via XState lifecycles and handles.

techsmyth replied:
To discuss regarding design + when / where to use xstate, which does have an overhead.

valentinyanakiev commented on src/domain/community/user/user.resolver.subscriptions.ts:
```diff
@@ -30,17 +42,39 @@ export class UserResolverSubscriptions {
+      AuthorizationPrivilege.UPDATE,
+      `subscribe to user message received events: ${user.displayName}`
+    );
+    return this.pubSub.asyncIterator(
```
This (and overall all events that should clearly be managed via lifecycle) should be managed via XState lifecycles and handles.

valentinyanakiev commented on src/domain/community/user/user.resolver.subscriptions.ts:
```diff
@@ -30,17 +42,39 @@ export class UserResolverSubscriptions {
   })
   roomNotificationReceived() {
-    return this.pubSub.asyncIterator(MATRIX_ROOM_JOINED);
+    return this.pubSub.asyncIterator(
```
This (and overall all events that should clearly be managed via lifecycle) should be managed via XState lifecycles and handles.

valentinyanakiev commented on src/domain/community/user/user.service.ts:
```diff
@@ -1,45 +1,69 @@
 @Injectable()
 export class UserService {
   replaceSpecialCharacters = require('replace-special-characters');
+  cacheOptions: CachingConfig = { ttl: 300 };
```
This can't be scaled horizontally (see comments above).

valentinyanakiev commented on src/services/platform/communication/communication.dto.delete.msg.community.ts:
```diff
@@ -0,0 +1,5 @@
+export class CommunicationDeleteMessageFromCommunityRoomInput {
+  sendingUserEmail!: string;
```
Extract e.g. BaseCommunicationMessage - you have a lot of copy-paste over the classes.

valentinyanakiev commented on src/services/platform/communication/communication.service.ts:
```diff
@@ -339,47 +403,62 @@ export class CommunicationService {
     return roomResult;
   }
 
   async convertMatrixRoomToCommunityRoom(
```
This and the method above look 90% identical, there could be a method extracted to remove code duplication.

techsmyth replied:
resolved in 458892759627d6dc13ec01ccb16f5b3c214c4fda

valentinyanakiev commented on src/services/platform/communication/communication.service.ts:
```diff
@@ -82,6 +84,50 @@ export class CommunicationService {
+    );
+  }
+
+  async deleteMsgCommunity(
```
Refactor this method and the method above - reuse the checks and extract the common code in a method.

techsmyth replied:
Done in 79c84d80a1b6867ddacea67831a48ce87b931566

valentinyanakiev commented on src/domain/community/community/community.resolver.mutations.ts:
```diff
@@ -275,4 +315,31 @@ export class CommunityResolverMutations {
+  })
+  @Profiling.api
+  async removeDiscussionMessageFromCommunity(
+    @Args('msgData') msgData: CommunityRemoveMessageInput,
```
This is exactly the same code as above but one call - extract the common logic.

techsmyth replied:
to discuss in wider context of authorization + duplication

valentinyanakiev commented on src/domain/community/community/community.service.ts:
```diff
@@ -456,4 +459,38 @@ export class CommunityService {
+    email: string,
+    msgData: CommunityRemoveMessageInput
+  ) {
+    await this.communicationService.ensureUserHasAccesToCommunityMessaging(
```
This method and the method above can be refactored to use common method for 90% of the logic.

valentinyanakiev commented on src/services/platform/matrix/agent/matrix.agent.service.ts:
```diff
@@ -225,4 +226,53 @@ export class MatrixAgentService {
+      )
+    );
+
+    // const response = await matrixAgent.matrixClient.sendEvent(
```
Are these comments a toDo or they can be removed?

techsmyth replied:
As discussed, todo above the method applies here

valentinyanakiev reviewed (changes requested):
- For now:
   - There is a lot of duplicate code that can be easily refactored. 
   - Method naming is not consistent with the application / PR. 
- For discussion:
   - [Sending events should be via lifecycle callbacks](https://app.zenhub.com/workspaces/alkemio-5ecb98b262ebd9f4aec4194c/issues/alkem-io/server/1388)
   - [Cache manager won't work in a production setup - we need Redis](https://app.zenhub.com/workspaces/alkemio-5ecb98b262ebd9f4aec4194c/issues/alkem-io/server/1387)

techsmyth committed:
Merge branch 'develop' into server-937

techsmyth committed:
updated names of methods / mutations

techsmyth committed:
refactored to reduce code duplication

techsmyth committed:
refactored code for converting matrix room to Alkemio rooms

Enderedne commented on src/domain/community/community/community.service.ts:
```diff
@@ -423,36 +426,70 @@ export class CommunityService {
   }
 
-  async sendDiscussionMessageToCommunity(
+  async sendMessageToCommunityDiscussions(
```
Should be sendMessageToCommunityDiscussions**Room** as in sendMessageToCommunityUpdatesRoom for consistency

techsmyth replied:
Addressed in 896940843be935910f4036fba30e554b1e1411ad - removed Room from all methods

Enderedne commented on src/services/platform/matrix/agent/matrix.agent.service.ts:
```diff
@@ -195,34 +196,85 @@ export class MatrixAgentService {
     return room.room_id;
   }
 
   async message(
```
We probably should rename this to sendMessage since we have deleteMessage and editMessage

techsmyth replied:
Good point - addressed in 896940843be935910f4036fba30e554b1e1411ad

Note: also removed a method in the same file that is not used - please check.

techsmyth committed:
address naming feedback

techsmyth committed:
missed files

techsmyth committed:
removed commented out method

techsmyth committed:
fixed property name as expected by matrix

techsmyth committed:
merged from develop; resolved conflicts

techsmyth committed:
bumped version
